cloud: add private link connection docs (#22192)
pingcap/docs: tidb-cloud/serverless-private-link-connection-to-self-hosted-kafka-in-alicloud.md

    1. Enter **Name**, for example, `Kafka VPC`.
    2. Select the region in which you want to set up the private link connection in TiDB Cloud.
    3. Select **Manually enter an IPv4 CIDR block**, and enter the IPv4 CIDR, for example, `10.0.0.0/16`.
    4. Create a vSwitch and configure IPv4 CIDR for each AZ in which you want to deploy Kafka brokers. For example:

        - broker-ap-southeast-1a vSwitch in `ap-southeast-1a`: 10.0.0.0/18
        - broker-ap-southeast-1b vSwitch in `ap-southeast-1b`: 10.0.64.0/18
        - broker-ap-southeast-1c vSwitch in `ap-southeast-1c`: 10.0.128.0/18
        - bastion vSwitch in `ap-southeast-1a`: 10.0.192.0/18

    5. Use the default values for other options. Click **Ok**.

3. On the VPC detail page, take note of the VPC ID, for example, `vpc-t4nfx2vcqazc862e9fg06`.

#### 2. Set up Kafka brokers

**2.1. Create a bastion node**

Go to the [ECS console](https://ecs.console.alibabacloud.com/home#/). Create the bastion node in the bastion vSwitch.

- **Network and Zone**: `Kafka VPC` and `bastion` vSwitch.
- **Instance and Image**: `ecs.t5-lc1m2.small` instance type and `Alibaba Cloud Linux` image.

> - **Network and Security Groups**: check `Assign Public IPv4 Address`.

Reviewer: ```suggestion
- **Network and Security Groups**: select `Assign Public IPv4 Address`.
```

---
title: Connect to Alibaba Cloud Self-Hosted Kafka via Private Link Connection
summary: Learn how to connect to an Alibaba Cloud self-hosted Kafka using an Alibaba Cloud Endpoint Service private link connection.
---

# Connect to Alibaba Cloud Self-Hosted Kafka via a Private Link Connection

The document describes how to connect to a self-hosted Kafka cluster in Alibaba Cloud, using Alibaba Cloud Endpoint Service private link connection.

The mechanism works as follows:

1. The private link connection connects to your endpoint service with bootstrap port, it returns broker external addresses defined in advertised.listeners.
2. The private link connection connects to your endpoint service with broker external addresses.
3. Endpoint service forwards requests to your load balancers.

> 4. Load balancers forward requests to different Kafka brokers based on the port mapping.

Reviewer 2: ```suggestion
1. The private link connection connects to your Alibaba Cloud endpoint service using the bootstrap port, which returns broker external addresses defined in `advertised.listeners`.
2. The private link connection connects to your endpoint service with broker external addresses.
3. The Alibaba Cloud endpoint service forwards requests to your load balancers.
4. Load balancers forward requests to the corresponding Kafka brokers based on port mapping.
```

For example, the port mapping is as follows:

| Broker External Address Port | Load Balancer Listener Port | Load Balancer Backend Server |
|----------------------------|------------------------------|-------------|
| 9093                       | 9093                        | broker-node1:39092|
| 9094                       | 9094                        | broker-node2:39092|
| 9095                       | 9095                        | broker-node3:39092|

## Prerequisites
   
- Ensure that you have a Kafka cluster or have the following permissions to set up one.

    - Manage ECS nodes
    - Manage VPC and vSwitch
    - Connect to ECS nodes to configure Kafka nodes

> - Ensure that you have the following authorization to set up a load balancer and endpoint service in your own Alibaba Cloud account.

Reviewer 2: ```suggestion
- Ensure that you have the following permissions to set up a load balancer and endpoint service in your Alibaba Cloud account.
```

    - Manage load balancer
    - Manage endpoint services

- {{{ .essential }}} information: confirm that your {{{ .essential }}} is active in Alibaba Cloud provider. Retrieve and save the following details for later use:

>     - Account ID

Reviewer 2: ```suggestion
    - Alibaba Cloud account ID
```

    - Availability Zones (AZ)

To view the the Alibaba Cloud account ID and available zones, do the following:

1. In the [TiDB Cloud console](https://tidbcloud.com), navigate to the cluster overview page of the TiDB cluster, and then click **Settings** > **Networking** in the left navigation pane.
2. On the **Private Link Connection For Dataflow**, click **Create Private Link Connection**.
3. You can find the Alibaba Cloud account ID and available zones information.

The following table shows an example of the deployment information.

| Information     | Value    | Note    | 
|--------|-----------------|---------------------------|
| Region    | `ap-southeast-1`   |  N/A |
| TiDB Cloud Alibaba Cloud Account | `<account_id>`     |    N/A  |
| AZ IDs                              | <ul><li>`ap-southeast-1a` </li><li>`ap-southeast-1b` </li><li> `ap-southeast-1c`</li></ul>  

> | Kafka Advertised Listener Pattern   | &lt;broker_id&gt;.unique_name.alicloud.plc.tidbcloud.com:&lt;port&gt;| `unique_name` is a placeholder. The actual value of `unique_name` will be replaced in [Step 4](#step-4-replace-the-unique-name-placeholder-in-kafka-configuration) |

Reviewer 2: ```suggestion
| TiDB Cloud Alibaba Cloud account | `<account_id>`     |    N/A  |
| AZ IDs                              | <ul><li>`ap-southeast-1a` </li><li>`ap-southeast-1b` </li><li> `ap-southeast-1c`</li></ul>  
| Kafka Advertised Listener Pattern   | &lt;broker_id&gt;.unique_name.alicloud.plc.tidbcloud.com:&lt;port&gt;| `unique_name` is a placeholder and will be replaced with the actual value in [Step 4](#step-4-replace-the-unique-name-placeholder-in-kafka-configuration) |
```

## Step 1. Set up a Kafka cluster

If you need to deploy a new cluster, follow the instructions in [Deploy a new Kafka cluster](#deploy-a-new-kafka-cluster).

If you need to expose an existing cluster, follow the instructions in [Reconfigure a running Kafka cluster](#reconfigure-a-running-kafka-cluster). 

### Deploy a new Kafka cluster

#### 1. Set up the Kafka VPC

The Kafka VPC requires the following:

- Three private vSwitches for brokers, one for each AZ. 
- One public vSwitches in any AZ with a bastion node that can connect to the internet and three private vSwitches, which makes it easy to set up the Kafka cluster. In a production environment, you might have your own bastion node that can connect to the Kafka VPC.

Take the following steps to create the Kafka VPC.

**1.1. Create the Kafka VPC**

1. Go to [Alibaba Cloud Console > VPC dashboard](https://vpc.console.alibabacloud.com/vpc), and switch to the region in which you want to deploy Kafka.

2. Click **Create VPC**. Fill in the information on the **VPC settings** page as follows.

> 

Reviewer 2: In this step, does the user need to select the region?

```

## Step 2. Expose the Kafka cluster as a private link service

### 1. Set up the load balancer

Create a network load balancer with four server groups with different ports. One server group is for bootstrap, and the others will map to different brokers.

1. bootstrap server group => 9092 => broker-node1:39092,broker-node2:39092,broker-node3:39092
2. broker server group 1  => 9093 => broker-node1:39092
3. broker server group 2  => 9094 => broker-node2:39092
4. broker server group 3  => 9095 => broker-node3:39092

If you have more broker role nodes, you need to add more mappings. Ensure that you have at least one node in the bootstrap target group. It is recommended to add three nodes, one for each AZ for resilience.

Do the following to set up the load balancer:

1. Go to [Server Groups](https://slb.console.alibabacloud.com/nlb/ap-southeast-1/server-groups) to create four server groups.

    - Bootstrap server group 

        - **Server Group Type**: select `Server`
        - **Server Group Name**: `bootstrap-server-group`
        - **VPC**: `Kafka VPC`
        - **Backend Server Protocol**: select `TCP`

>         - **Backend servers**:  Click the created server group to add backend servers, add `broker-node1:39092`, `broker-node2:39092`, `broker-node3:39092`

Reviewer 2: ```suggestion
        - **Backend servers**: click the created server group and add backend servers, including `broker-node1:39092`, `broker-node2:39092`, and `broker-node3:39092`
```

    - Broker server group 1

        - **Server Group Type**: select `Server`
        - **Server Group Name**: `broker-server-group-1`
        - **VPC**: `Kafka VPC`
        - **Backend Server Protocol**: select `TCP`
        - **Backend servers**:  Click the created server group to add backend servers, add `broker-node1:39092`

    - Broker server group 2

        - **Server Group Type**: select `Server`
        - **Server Group Name**: `broker-server-group-2`
        - **VPC**: `Kafka VPC`
        - **Backend Server Protocol**: select `TCP`

>         - **Backend servers**:  Click the created server group to add backend servers, add `broker-node2:39092`

Reviewer 2: ```suggestion
        - **Backend servers**: click the created server group and add the backend server `broker-node2:39092`
```
